File: src/main/java/me/pcasaes/hexoids/configuration/ClientEventProducerProvider.java
package me.pcasaes.hexoids.configuration;

import me.pcasaes.hexoids.entrypoints.web.ClientBroadcaster;
import me.pcasaes.hexoids.infrastructure.producer.ClientEventProducer;
import pcasaes.hexoids.proto.Dto;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;
import javax.inject.Inject;
import javax.inject.Singleton;

@ApplicationScoped
public class ClientEventProducerProvider {

    private final ClientBroadcaster clientBroadcaster;

    @Inject
    public ClientEventProducerProvider(ClientBroadcaster clientBroadcaster) {
        this.clientBroadcaster = clientBroadcaster;
    }

    @Produces
    @Singleton
    public ClientEventProducer getClientBroadcaster() {
        return new ClientEventProducer() {
            @Override
            public boolean isEnabled() {
                return clientBroadcaster.isEnabled();
            }

            @Override
            public void accept(Dto dto) {
                clientBroadcaster.accept(dto);
            }

            @Override
            public String getName() {
                return clientBroadcaster.getName();
            }
        };
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/configuration/ClientEventProducerProvider.kt
package me.pcasaes.hexoids.configuration

import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.inject.Produces
import jakarta.inject.Inject
import jakarta.inject.Singleton
import me.pcasaes.hexoids.entrypoints.web.ClientBroadcaster
import me.pcasaes.hexoids.infrastructure.producer.ClientEventProducer
import pcasaes.hexoids.proto.Dto

@ApplicationScoped
class ClientEventProducerProvider @Inject constructor(
    private val clientBroadcaster: ClientBroadcaster
) {
    @Produces
    @Singleton
    fun getClientBroadcaster(): ClientEventProducer {
        return object : ClientEventProducer {
            override fun isEnabled(): Boolean {
                return clientBroadcaster.isEnabled()
            }

            override fun accept(dto: Dto?) {
                clientBroadcaster.accept(dto)
            }
        }
    }
}
